test/test_cuda.py

```diff
@@ -5613,6 +5613,129 @@ def my_function(pool):
             s = p.snapshot()
             self.assertEqual(len(s), 1, "Expected to have a single segment")
 
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs"
+    )
+    def test_graph_capture_reclaim_2_streams(self):
+        torch.cuda.memory._set_allocator_settings(
+            "graph_capture_record_stream_reuse:True"
+        )
+        torch.cuda.empty_cache()
+        s1 = torch.cuda.Stream()
+        s2 = torch.cuda.Stream()
+        g = torch.cuda.CUDAGraph(keep_graph=True)
+
+        torch.cuda.synchronize()
+
+        with torch.cuda.stream(s1):
+            g.capture_begin()
+            unrelated = torch.rand(8, device="cuda")
+
+            data1 = torch.rand(8, device="cuda")
+            data1_ptr = data1.data_ptr()
+
+            s2.wait_stream(s1)
+            with torch.cuda.stream(s2):
+                # Mark data1 as used on stream 2
+                data1.record_stream(s2)
+
+            del data1
+
+            s1.wait_stream(s2)
+
+            # Insert a dummy operation to ensure the graph is joined after this point
+            unrelated.fill_(1.0)
+            # The next allocation should reuse data1's memory if graph_capture_record_stream_reuse is enabled
+            data2 = torch.rand(8, device="cuda")
+            data2_ptr = data2.data_ptr()
+
+            g.capture_end()
+
+        torch.cuda.synchronize()
+        self.assertTrue(data1_ptr == data2_ptr)
+
+        torch.cuda.memory._set_allocator_settings(
+            "graph_capture_record_stream_reuse:False"
+        )
+
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs"
+    )
+    def test_graph_capture_reclaim_4_streams(self):
+        torch.cuda.memory._set_allocator_settings(
+            "graph_capture_record_stream_reuse:True"
+        )
+
+        torch.cuda.empty_cache()
+        s1 = torch.cuda.Stream()
+        s2 = torch.cuda.Stream()
+
+        s3 = torch.cuda.Stream()
+        s4 = torch.cuda.Stream()
+        g = torch.cuda.CUDAGraph(keep_graph=True)
+
+        torch.cuda.synchronize()
+
+        with torch.cuda.stream(s1):
+            g.capture_begin()
+            unrelated = torch.rand(8, device="cuda")
+
+            data1 = torch.ones(8, device="cuda")
+            data1_ptr = data1.data_ptr()
+
+            s2.wait_stream(s1)
+            with torch.cuda.stream(s2):
+                # Mark data1 as used on stream 2
```

Comment:
can you actually put some work here? Right now the forks in the graph are empty (which maybe fine, but is still somewhat confusing for the test)